---
title: "Bulk RNA sequencing analysis: Human, murine and iPSC-derived yolk sac"
author: "SChear"
date: "2024-03-08"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

```




```{r,results='hide',fig.keep='all', warning=FALSE,message=FALSE}
library(DESeq2)
library(ggplot2)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(ComplexHeatmap)
library(clusterProfiler)
library(EnhancedVolcano)
library(fgsea)
library(magrittr)
library(tidyverse)
library(vsn)
library(pheatmap)
library(RColorBrewer)
library(edgeR)
library(matrixStats)
library(circlize)
library(tximport)
library(ggvenn)
library(grid)
library(ggplot2)

```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
setwd("/Users/sueannechear/Bioinformatics/yolk_sac/analysis ")
counts<-read.csv("cleancount.csv",header=T, sep=",")
rownames(counts)  <- counts[,1]
counts<- counts[ -c(1) ]
head(counts,3)

```

Samples used in this analysis:

```{r}
df<-data.frame("Abbreviation"=c("MYS","HYS","iYS"),
               "Sample"=c("Mouse yolk sac","HUman yolk sac","iYolk sac"),
               "Source"=c("PRJEB18767 ","PRJEB18767 ","Da's culture"), "Replicates"=c(3,3,3))
print(df)
```


Create a DESeq dataset : 9 samples; 19267 genes
```{r,fig.keep='all', warning=FALSE,message=FALSE}
condition<-factor(c("MYS","MYS","MYS","iYS","iYS",
                    "iYS","HYS","HYS","HYS"))

sample<-factor(colnames(counts))

coldata<-data.frame(sample,condition)

dds<-DESeqDataSetFromMatrix(countData = counts,
                            colData = coldata,
                            design=~condition)

dds
```

We retain genes with at least 1 count per million (CPM) in at least two samples.
Genes remained after filtering: 15007
```{r,fig.keep='all', warning=FALSE,message=FALSE}
dds = dds[ rowSums(edgeR::cpm(counts(dds)) > 1)>=2, ]

nrow(dds)
```

QC for dispersion of variability in data. A normal expected plot is that the fitted line of estimates is high on the y axis for lower counts but taper off to plateau towards genes with
higher counts; ideally we want fitted line trend below 1, indicating the data is a good fit
for the DESeq model.
```{r, results='hide',fig.keep='all', warning=FALSE,message=FALSE}
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
plotDispEsts(dds)
```

Transform data using VST method before PCA.

```{r, results='hide', warning=FALSE,message=FALSE}
vstdata<-vst(dds,blind=F)

meanSdPlot(assay(vstdata), ranks=FALSE)
```

PCA: to examine variation between samples. A principal component analysis (PCA) suggested that samples cluster according to species but not to tissue sites (data are based on ortholog determined with Biomart),which indicates that the phylogenetic divergence is stronger than site differences. This could be a consequence of analysis based on the limited 15,007 two-way orthologs that can be mapped 1:1 between Human-Mouse.

```{r,results='hide', warning=FALSE, message=FALSE}

plotPCA(vstdata,intgroup="condition")
```



```{r, results='hide',warning=FALSE,message=FALSE}

#check pc3 and pc4
plotPCA <- function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = " : "))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], PC3 = pca$x[, 3], PC4 = pca$x[, 4], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[1:2]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC3", y = "PC4", color = "group")) + 
        geom_point(size = 3) + xlab(paste0("PC3: ", round(percentVar[1] * 
        100), "% variance")) + ylab(paste0("PC4: ", round(percentVar[2] * 
        100), "% variance")) + coord_fixed()
}

#print(plotPCA(vstdata,intgroup="condition"))
```


```{r}
ntop <- 500
rv <- rowVars(assay(vstdata))
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
mat <- t( assay(vstdata)[select, ] )

pca<-prcomp(mat)
pca <- as.data.frame(pca$x)

```



Which genes impact the most in PC1?

```{r}
getLoadings = function(dds){
  
  mat<-assay(vstdata)
  pca = prcomp(t(mat), retx = TRUE)
  
  return(pca$rotation)
}

loadings_vstdata = getLoadings(vstdata) %>% as.data.frame()
# Annotating gene names
loadings_vstdata$symbol = mapIds(org.Hs.eg.db,
                              keys=rownames(loadings_vstdata),
                              column="SYMBOL",
                              keytype="ENSEMBL",
                              multiVals="first")

# show the top 50 genes from PC1
loadings_vstdata %>% 
  # select only the PCs we are interested in
  dplyr::select(symbol, PC1) %>%
  # convert to "long" format
  pivot_longer(cols = "PC1", names_to = "PC1", values_to = "loadings") %>% 
  # for PC1
  group_by(PC1) %>% 
  # arrange by descending order
  arrange(desc(abs(loadings))) %>% 
  # take the 10 top rows
  slice(1:50) %>%
  pull(symbol)

```
Which genes impact the most in PC2?

```{r}

# show the top 20 genes from PC2
loadings_vstdata %>% 
  # select only the PCs we are interested in
  dplyr::select(symbol, PC2) %>%
  # convert to "long" format
  pivot_longer(cols = "PC2", names_to = "PC2", values_to = "loadings") %>% 
  # for PC2
  group_by(PC2) %>% 
  # arrange by descending order
  arrange(desc(abs(loadings))) %>% 
  # take the 10 top rows
  slice(1:50) %>%
  pull(symbol)
```


Cluster dendrogram: 
```{r,results='hide', fig.keep='all',warning=FALSE,message=FALSE}
rv <- rowVars(assay(vstdata))
o <- order(rv,decreasing=TRUE)
dists <- dist(t(assay(vstdata)[head(o,500),]))
hc <- hclust(dists)
plot(hc, labels=vstdata$sample)
```


Correlation matrix heat map of transcript expression across all samples. Euclidean distance
is used as the similarity measure and clustering samples based on the 'complete' method.
The lower the numbers, the stronger the correlation between samples. 

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}

sampleDists<-dist(t(assay(vstdata)))
sampleDistMatrix<-as.matrix(sampleDists)
colnames(sampleDistMatrix)

colors<-colorRampPalette(rev(brewer.pal(9,"Blues")))(255)

pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists, display_numbers = TRUE,
         clustering_distance_cols=sampleDists, col=colors, fontsize_number=10,cellnote = mat,cellnote_fontface = "bold")
```

The heatmap becomes more interesting if we do not look at absolute expression strength but rather at the amount by which each gene deviates in a specific sample from the gene’s average across all samples. Hence, we center each genes’ values across samples, and plot a heatmap (figure below).  

```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
#get the indices of the top variable genes
topVarGenes <- head(order(rowVars(assay(vstdata)), decreasing = TRUE), 100)

#subset the data matrix to include only the top variable genes
mat  <- assay(vstdata)[ topVarGenes, ]

#center the data
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vstdata))

#map ensembl IDs to gene symbols
symbols <- mapIds(org.Hs.eg.db, keys = rownames(mat), column = "SYMBOL", keytype = "ENSEMBL")

#assign gene symbols as row names
rownames(mat) <- symbols

#create a heatmap with hierarchical clustering
#heatmap_result <- pheatmap(mat, annotation_col = anno, fontsize_row=5)


```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
# Perform hierarchical clustering separately
hc_rows <- hclust(dist(mat), method = "complete")

# Create a heatmap with hierarchical clustering
#heatmap_result <- pheatmap(mat, annotation_col = anno, fontsize_row = 5, clustering_distance_rows = "correlation")

heatmap_result <- pheatmap(mat, fontsize_row = 5, clustering_distance_rows = "correlation")
# Extract the cluster assignments for the rows
cluster_assignments <- cutree(hc_rows, k = 3)

# Print or use the cluster assignments as needed
print(cluster_assignments)

# Create a data frame with gene symbols and cluster assignments
gene_cluster_df <- data.frame(
  GeneSymbol = names(cluster_assignments),
  ClusterAssignment = cluster_assignments
)

# Order the data frame by cluster assignments
gene_cluster_df <- gene_cluster_df[order(gene_cluster_df$ClusterAssignment, gene_cluster_df$GeneSymbol), ]

# Print or use the data frame as needed
print(gene_cluster_df)

write.csv(gene_cluster_df, "genecluster.csv",row.names=F)
```

Pathways enriched in cluster 1 :these are genes upregulated in mouse yolk sac.
```{r,fig.keep='all', results='hide', warning=FALSE,message=FALSE}
# Select genes in Cluster 1
genes_in_cluster1 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 1]

print(genes_in_cluster1)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster1, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster1, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

Pathways enriched in cluster 2; these are genes upregulated in human yolk sac associated with lipid transport and cholesterol.During organogenesis, the embryo is reliant on maternal sources of cholesterol until its liver is sufficiently mature for synthesis.
```{r,fig.keep='all', results='hide', fig.width=10, fig.height=10,warning=FALSE,message=FALSE}
# Select genes in Cluster 2
genes_in_cluster2 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 2]

print(genes_in_cluster2)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster2, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster2, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

Pathways enriched in cluster 3 ; these are genes upregulated in both human and iyolk sac.
```{r,fig.keep='all', results='hide', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Select genes in Cluster 3
genes_in_cluster3 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 3]

print(genes_in_cluster3)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster3, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster3, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

```{r,fig.keep='all', warning=FALSE,message=FALSE}
#combine gene cluster list into dataframe
cluster_list<-list(genes_in_cluster1,genes_in_cluster2,genes_in_cluster3)

cluster1<-as.data.frame(genes_in_cluster1)%>%mutate(cluster=1)%>%rename(genes = genes_in_cluster1)
cluster2<-as.data.frame(genes_in_cluster2)%>%mutate(cluster=2)%>%rename(genes = genes_in_cluster2)
cluster3<-as.data.frame(genes_in_cluster3)%>%mutate(cluster=3)%>%rename(genes = genes_in_cluster3)

combined_cluster<-rbind(cluster1,cluster2,cluster3)

write.csv(combined_cluster, "combined_cluster.csv",row.names=F,sep='\t')
```



## Differential expression of gene analysis:  
### HYS vs iYS. (p adjusted value=0.05)

```{r,fig.keep='all', warning=FALSE,message=FALSE}
pds<-dds

pds$condition<-relevel(pds$condition, ref="iYS")

pds<-DESeq(pds)

res = results(pds, contrast=c("condition","HYS","iYS"), alpha=0.05)

summary(res)

```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
sigs<-na.omit(res)

sigs.df<-as.data.frame(sigs)

columns(org.Hs.eg.db)

sigs.df$symbol<-mapIds(org.Hs.eg.db, keys=rownames(sigs.df), keytype = "ENSEMBL",column="SYMBOL")

sigs.df<-sigs.df%>%filter(!str_detect(symbol,'NA'))


```


Retrieve gene ontology terms associated with upregulated genes in HYS vs iYS samples. 

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange >5,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```


### MYS vs iYS (p adjusted value=0.05). 

```{r,fig.keep='all', warning=FALSE,message=FALSE}
pds<-dds

pds$condition<-relevel(pds$condition, ref="iYS")

pds<-DESeq(pds)

res = results(pds, contrast=c("condition","MYS","iYS"), alpha=0.05)

summary(res)

```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
sigs<-na.omit(res)

sigs.df<-as.data.frame(sigs)

columns(org.Hs.eg.db)

sigs.df$symbol<-mapIds(org.Hs.eg.db, keys=rownames(sigs.df), keytype = "ENSEMBL",column="SYMBOL")

sigs.df<-sigs.df%>%filter(!str_detect(symbol,'NA'))


```


Retrieve gene ontology terms associated with upregulated genes in MYS vs iYS samples
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange >5,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```

Retrieve gene ontology terms associated with downregulated genes in MYS vs iYS samples
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange < -2,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```

### HYS vs MYS (p adjusted value=0.05). 

```{r,fig.keep='all', warning=FALSE,message=FALSE}
pds<-dds

pds$condition<-relevel(pds$condition, ref="MYS")

pds<-DESeq(pds)

res = results(pds, contrast=c("condition","HYS","MYS"), alpha=0.05)

summary(res)

```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
sigs<-na.omit(res)

sigs.df<-as.data.frame(sigs)

columns(org.Hs.eg.db)

sigs.df$symbol<-mapIds(org.Hs.eg.db, keys=rownames(sigs.df), keytype = "ENSEMBL",column="SYMBOL")

sigs.df<-sigs.df%>%filter(!str_detect(symbol,'NA'))


```


Retrieve gene ontology terms associated with upregulated genes in HYS vs MYS samples
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange >5,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```

Retrieve gene ontology terms associated with downregulated genes in HYS vs MYS samples
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange < -2,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```

## Gene level abundance in each species.  

Transcript counts TPM were summarized to gene level abundance and averaged across replicates.
Top 400 most abundant genes in each species were used for pathway enrichment analysis.

### Pathways enriched in top 400 genes in iYolk Sac. 

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Assuming rep1, rep2, rep3 are your replicate abundance files
setwd("/Users/sueannechear/Bioinformatics/yolk_sac/data2/abundance")

rep1 <- read.table("iMAC1_abundance.tsv", header = TRUE, sep = "\t")
rep2 <- read.table("iMAC2_abundance.tsv", header = TRUE, sep = "\t")
rep3 <- read.table("iMAC3_abundance.tsv", header = TRUE, sep = "\t")


library(tximport)

# Assuming tx2gene is your transcript-to-gene mapping table
tx2gene <- read.table("tx2gene_EnsDb_Hsapiens.txt", header = TRUE)


# Run tximport with the averaged abundances
txi <- tximport(files = c("iMAC1_abundance.tsv", "iMAC2_abundance.tsv", "iMAC3_abundance.tsv"),
                type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE, ignoreTxVersion = TRUE)

head(txi$counts)

# Get the current column names (sample names)
current_sample_names <- colnames(txi$counts)

# Create a new set of sample names
new_sample_names <- c("iMAC1", "iMAC2", "iMAC3")

# Rename the columns in counts, abundance, or other relevant slots
colnames(txi$counts) <- new_sample_names
colnames(txi$abundance) <- new_sample_names
head(txi$counts)

# Extract counts matrix
counts_matrix <- txi$counts

# Calculate the average abundance across replicates
average_abundance <- rowMeans(counts_matrix)


# Add the average abundance as a new column in the counts_matrix data frame
counts_matrix <- as.data.frame(counts_matrix)  # Convert to data frame if not already
counts_matrix <- counts_matrix %>%
  mutate(average_abundance = average_abundance)

# Arrange the genes based on total_counts in descending order
gene_sorted <- counts_matrix %>%
  arrange(desc(average_abundance))

# Select the top 400 genes
top_400_genes <- head(gene_sorted, 400)

# Display the top 400 genes
print(top_400_genes)

# Map gene symbols to Entrez IDs
entrez_ids <- mapIds(org.Hs.eg.db, keys=rownames(top_400_genes), keytype = "ENSEMBL",column="SYMBOL")

# Remove NA values from entrez_ids (genes without matching Entrez IDs)
entrez_ids <- entrez_ids[!is.na(entrez_ids)]

#convert entrez_ids to dataframe
top400_iYS<-as.data.frame(entrez_ids)

# Enrichment analysis for biological processes (BP)
GO_results <- enrichGO(gene = entrez_ids, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
GO_results_df <- as.data.frame(GO_results)

# Print the top GO terms
print(GO_results_df)


plot(barplot(GO_results, showCategory = 15))

```

### Pathways enriched in top 400 genes in human yolk sac.  

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Assuming rep1, rep2, rep3 are your replicate abundance files
setwd("/Users/sueannechear/Bioinformatics/yolk_sac/data2/abundance")

rep1 <- read.table("YS7_H_abundance.tsv", header = TRUE, sep = "\t")
rep2 <- read.table("YS9_H_abundance.tsv", header = TRUE, sep = "\t")
rep3 <- read.table("YS10_H_abundance.tsv", header = TRUE, sep = "\t")


# Run tximport with the averaged abundances
txj <- tximport(files = c("YS7_H_abundance.tsv", "YS9_H_abundance.tsv", "YS10_H_abundance.tsv"),
                type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE, ignoreTxVersion = TRUE)

head(txj$counts)

# Get the current column names (sample names)
current_sample_names <- colnames(txj$counts)

# Create a new set of sample names
new_sample_names <- c("HYS1", "HYS2", "HYS3")

# Rename the columns in counts, abundance, or other relevant slots
colnames(txj$counts) <- new_sample_names
colnames(txj$abundance) <- new_sample_names
head(txj$counts)

# Extract counts matrix
counts_matrix <- txj$counts

# Calculate the average abundance across replicates
average_abundance <- rowMeans(counts_matrix)


# Add the average abundance as a new column in the counts_matrix data frame
counts_matrix2 <- as.data.frame(counts_matrix)  # Convert to data frame if not already
counts_matrix2 <- counts_matrix2 %>%
  mutate(average_abundance = average_abundance)

# Arrange the genes based on total_counts in descending order
gene_sorted2 <- counts_matrix2 %>%
  arrange(desc(average_abundance))

# Select the top 400 genes
top_400_genes2 <- head(gene_sorted2, 400)

# Display the top 400 genes
print(top_400_genes2)

# Map gene symbols to Entrez IDs
entrez_ids <- mapIds(org.Hs.eg.db, keys=rownames(top_400_genes2), keytype = "ENSEMBL",column="SYMBOL")

#convert entrez_ids to data frame
top400_HYS<-as.data.frame(entrez_ids)

# Remove NA values from entrez_ids (genes without matching Entrez IDs)
entrez_ids <- entrez_ids[!is.na(entrez_ids)]

# Enrichment analysis for biological processes (BP)
GO_results <- enrichGO(gene = entrez_ids, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
GO_results_df <- as.data.frame(GO_results)

# Print the top GO terms
print(GO_results_df)


plot(barplot(GO_results, showCategory = 15))

```

### Pathways enriched in top 400 genes in mouse yolk sac.  

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Assuming rep1, rep2, rep3 are your replicate abundance files
setwd("/Users/sueannechear/Bioinformatics/yolk_sac/data2/abundance")

rep1 <- read.table("YS3_M_abundance.tsv", header = TRUE, sep = "\t")
rep2 <- read.table("YS10_M_abundance.tsv", header = TRUE, sep = "\t")
rep3 <- read.table("YS11_M_abundance.tsv", header = TRUE, sep = "\t")

# Assuming tx2gene is your transcript-to-gene mapping table
tx2Mgene <- read.table("tx2gene_EnsDb_MMusculus.txt", header = TRUE)

# Run tximport with the averaged abundances
txm <- tximport(files = c("YS3_M_abundance.tsv", "YS10_M_abundance.tsv", "YS11_M_abundance.tsv"),
                type = "kallisto", tx2gene = tx2Mgene, ignoreAfterBar = TRUE, ignoreTxVersion = TRUE)

head(txm$counts)

# Get the current column names (sample names)
current_sample_names <- colnames(txm$counts)

# Create a new set of sample names
new_sample_names <- c("MYS1", "MYS2", "MYS3")

# Rename the columns in counts, abundance, or other relevant slots
colnames(txm$counts) <- new_sample_names
colnames(txm$abundance) <- new_sample_names
head(txm$counts)

# Extract counts matrix
counts_matrix <- txm$counts

# Calculate the average abundance across replicates
average_abundance <- rowMeans(counts_matrix)


# Add the average abundance as a new column in the counts_matrix data frame
counts_matrix2 <- as.data.frame(counts_matrix)  # Convert to data frame if not already
counts_matrix2 <- counts_matrix2 %>%
  mutate(average_abundance = average_abundance)

# Arrange the genes based on total_counts in descending order
gene_sorted2 <- counts_matrix2 %>%
  arrange(desc(average_abundance))

# Select the top 400 genes
top_400_genes2 <- head(gene_sorted2, 400)

# Display the top 400 genes
print(top_400_genes2)

# Map gene symbols to Entrez IDs
entrez_ids <- mapIds(org.Mm.eg.db, keys=rownames(top_400_genes2), keytype="ENSEMBL",column="SYMBOL")

#convert entrez_ids to data frame
top400_MYS<-as.data.frame(entrez_ids)

# Remove NA values from entrez_ids (genes without matching Entrez IDs)
entrez_ids <- entrez_ids[!is.na(entrez_ids)]

# Enrichment analysis for biological processes (BP)
GO_results <- enrichGO(gene = entrez_ids, OrgDb = org.Mm.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
GO_results_df <- as.data.frame(GO_results)

# Print the top GO terms
print(GO_results_df)


plot(barplot(GO_results, showCategory = 15))

```

### Overlapping genes in top 400 gene set between species. 
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Extract gene vectors from data frames
genes1 <- top400_HYS$entrez_ids
genes2 <- top400_iYS$entrez_ids
genes3 <- top400_MYS$entrez_ids

# Remove NAs from gene vectors
genes1 <- genes1[complete.cases(genes1)]
genes2 <- genes2[complete.cases(genes2)]
genes3 <- genes3[complete.cases(genes3)]


# Create a list of gene vectors
gene_lists <- list(HYS = genes1, iYS = genes2, MYS = genes3)

# Create a Venn diagram using ggvenn
venn <- ggvenn(gene_lists)

# Customize colors for each category
venn_plot <- venn + scale_fill_manual(values = c("red", "green", "blue"))

# Display the Venn diagram
print(venn_plot)


# Find the overlapping genes between 3 clusters
overlapping_genes <- Reduce(intersect, gene_lists)

print(overlapping_genes)

# Find overlapping genes between HYS and MYS
overlapping_genes_HYS_MYS <- intersect(gene_lists$HYS, gene_lists$MYS)

print(overlapping_genes_HYS_MYS)

# Find overlapping genes between HYS and iYS
overlapping_genes_HYS_iYS <- intersect(gene_lists$HYS, gene_lists$iYS)

print(overlapping_genes_HYS_iYS)

# Find overlapping genes between MYS and iYS
overlapping_genes_MYS_iYS <- intersect(gene_lists$MYS, gene_lists$iYS)

print(overlapping_genes_MYS_iYS)

```

## Find pathways enriched in overlapping genes. 

### HYS_iYS
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Enrichment analysis for biological processes (BP)
GO_HYS_iYS <- enrichGO(gene = overlapping_genes_HYS_iYS, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
as.data.frame(GO_HYS_iYS)

plot(barplot(GO_HYS_iYS, showCategory = 15,cex.names = 0.5))

```


### HYS_MYS
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Enrichment analysis for biological processes (BP)
GO_HYS_MYS <- enrichGO(gene = overlapping_genes_HYS_MYS, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
as.data.frame(GO_HYS_MYS)

plot(barplot(GO_HYS_MYS, showCategory = 15,cex.names = 0.5))

```

### iYS_MYS
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
# Enrichment analysis for biological processes (BP)
GO_MYS_iYS <- enrichGO(gene = overlapping_genes_MYS_iYS, OrgDb = org.Hs.eg.db, keyType = "SYMBOL", ont = "BP")

# Convert the results to a data frame
as.data.frame(GO_MYS_iYS)

plot(barplot(GO_MYS_iYS, showCategory = 15,cex.names = 0.5))

```

## Comparing data between HYS and iYS only. 

Load count matrix for HYS and iYS
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}

counts2<-read.csv("mergedH_iYScount.csv",header=T, sep=",")
rownames(counts2)  <- counts2[,1]
counts2<- counts2[ -c(1) ]
head(counts2,3)
counts2<-na.omit(counts2)
```

Create a DESeq dataset : 6 samples; 62,753 genes
```{r,fig.keep='all', warning=FALSE,message=FALSE}
condition<-factor(c("iYS","iYS","iYS","HYS","HYS","HYS"))

sample<-factor(colnames(counts2))

coldata<-data.frame(sample,condition)

dds<-DESeqDataSetFromMatrix(countData = counts2,
                            colData = coldata,
                            design=~condition)

dds
```

We retain genes with at least 1 count per million (CPM) in at least two samples.
Genes remained after filtering: 17,457
```{r,fig.keep='all', warning=FALSE,message=FALSE}
dds = dds[ rowSums(edgeR::cpm(counts(dds)) > 1)>=2, ]

nrow(dds)
```

QC for dispersion of variability in data. 
```{r, results='hide',fig.keep='all', warning=FALSE,message=FALSE}
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
plotDispEsts(dds)
```

Transform data using VST method before PCA. 
```{r, results='hide', warning=FALSE,message=FALSE}
vstdata<-vst(dds,blind=F)

meanSdPlot(assay(vstdata), ranks=FALSE)
```

PCA: to examine variation between samples. 
```{r,results='hide', warning=FALSE,message=FALSE}

plotPCA(vstdata,intgroup="condition")
```

```{r, results='hide',warning=FALSE,message=FALSE}

#check pc3 and pc4
plotPCA <- function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = " : "))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], PC3 = pca$x[, 3], PC4 = pca$x[, 4], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[1:2]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC3", y = "PC4", color = "group")) + 
        geom_point(size = 3) + xlab(paste0("PC3: ", round(percentVar[1] * 
        100), "% variance")) + ylab(paste0("PC4: ", round(percentVar[2] * 
        100), "% variance")) + coord_fixed()
}

#print(plotPCA(vstdata,intgroup="condition"))
```

```{r}
ntop <- 500
rv <- rowVars(assay(vstdata))
select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
mat <- t( assay(vstdata)[select, ] )

pca<-prcomp(mat)
pca <- as.data.frame(pca$x)

```



Which genes impact the most in PC1?

```{r}
getLoadings = function(dds){
  
  mat<-assay(vstdata)
  pca = prcomp(t(mat), retx = TRUE)
  
  return(pca$rotation)
}

loadings_vstdata = getLoadings(vstdata) %>% as.data.frame()
# Annotating gene names
loadings_vstdata$symbol = mapIds(org.Hs.eg.db,
                              keys=rownames(loadings_vstdata),
                              column="SYMBOL",
                              keytype="ENSEMBL",
                              multiVals="first")

# show the top 50 genes from PC1
loadings_vstdata %>% 
  # select only the PCs we are interested in
  dplyr::select(symbol, PC1) %>%
  # convert to "long" format
  pivot_longer(cols = "PC1", names_to = "PC1", values_to = "loadings") %>% 
  # for PC1
  group_by(PC1) %>% 
  # arrange by descending order
  arrange(desc(abs(loadings))) %>% 
  # take the 10 top rows
  slice(1:50) %>%
  pull(symbol)

```

Which genes impact the most in PC2?

```{r}

# show the top 20 genes from PC2
loadings_vstdata %>% 
  # select only the PCs we are interested in
  dplyr::select(symbol, PC2) %>%
  # convert to "long" format
  pivot_longer(cols = "PC2", names_to = "PC2", values_to = "loadings") %>% 
  # for PC2
  group_by(PC2) %>% 
  # arrange by descending order
  arrange(desc(abs(loadings))) %>% 
  # take the 10 top rows
  slice(1:50) %>%
  pull(symbol)
```


Cluster dendrogram:  

```{r,results='hide', fig.keep='all',warning=FALSE,message=FALSE}
rv <- rowVars(assay(vstdata))
o <- order(rv,decreasing=TRUE)
dists <- dist(t(assay(vstdata)[head(o,500),]))
hc <- hclust(dists)
plot(hc, labels=vstdata$sample)
```


Correlation matrix heat map of transcript expression across all samples. 

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}

sampleDists<-dist(t(assay(vstdata)))
sampleDistMatrix<-as.matrix(sampleDists)
colnames(sampleDistMatrix)

colors<-colorRampPalette(rev(brewer.pal(9,"Blues")))(255)

pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists, display_numbers = TRUE,
         clustering_distance_cols=sampleDists, col=colors, fontsize_number=10,cellnote = mat,cellnote_fontface = "bold")
```

### Heatmap to visualize clustering. 

```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
#get the indices of the top variable genes
topVarGenes <- head(order(rowVars(assay(vstdata)), decreasing = TRUE), 100)

#subset the data matrix to include only the top variable genes
mat  <- assay(vstdata)[ topVarGenes, ]

#center the data
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vstdata))

#map ensembl IDs to gene symbols
symbols <- mapIds(org.Hs.eg.db, keys = rownames(mat), column = "SYMBOL", keytype = "ENSEMBL")

#assign gene symbols as row names
rownames(mat) <- symbols

#create a heatmap with hierarchical clustering
#heatmap_result <- pheatmap(mat, annotation_col = anno, fontsize_row=5)


```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
# Perform hierarchical clustering separately
hc_rows <- hclust(dist(mat), method = "complete")

# Create a heatmap with hierarchical clustering
#heatmap_result <- pheatmap(mat, annotation_col = anno, fontsize_row = 5, clustering_distance_rows = "correlation")

heatmap_result <- pheatmap(mat, fontsize_row = 5, clustering_distance_rows = "correlation")
# Extract the cluster assignments for the rows
cluster_assignments <- cutree(hc_rows, k = 3)

# Print or use the cluster assignments as needed
print(cluster_assignments)

# Create a data frame with gene symbols and cluster assignments
gene_cluster_df <- data.frame(
  GeneSymbol = names(cluster_assignments),
  ClusterAssignment = cluster_assignments
)

# Order the data frame by cluster assignments
gene_cluster_df <- gene_cluster_df[order(gene_cluster_df$ClusterAssignment, gene_cluster_df$GeneSymbol), ]

# Print or use the data frame as needed
print(gene_cluster_df)

write.csv(gene_cluster_df, "genecluster2.csv",row.names=F)
```

Pathways enriched in cluster 1 :these are genes upregulated in human yolk sac.  

```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
# Select genes in Cluster 1
genes_in_cluster1 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 1]

print(genes_in_cluster1)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster1, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster1, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

Pathways enriched in cluster 2; these are genes upregulated in iYS_1 and iYS_2.  

```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
# Select genes in Cluster 2
genes_in_cluster2 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 2]

print(genes_in_cluster2)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster2, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster2, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

Pathways enriched in cluster 3 ; these are genes upregulated in iYS_3.  

```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
# Select genes in Cluster 3
genes_in_cluster3 <- gene_cluster_df$GeneSymbol[gene_cluster_df$ClusterAssignment == 3]

print(genes_in_cluster3)

entrez_ids <- mapIds(org.Hs.eg.db, keys = genes_in_cluster3, keytype = "SYMBOL", column = "ENTREZID")

GO_results <- enrichGO(gene = genes_in_cluster3, OrgDb = "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15,cex.names = 0.5))
```

```{r,fig.keep='all', warning=FALSE,message=FALSE}
#combine gene cluster list into dataframe
cluster_list<-list(genes_in_cluster1,genes_in_cluster2,genes_in_cluster3)

cluster1<-as.data.frame(genes_in_cluster1)%>%mutate(cluster=1)%>%rename(genes = genes_in_cluster1)
cluster2<-as.data.frame(genes_in_cluster2)%>%mutate(cluster=2)%>%rename(genes = genes_in_cluster2)
cluster3<-as.data.frame(genes_in_cluster3)%>%mutate(cluster=3)%>%rename(genes = genes_in_cluster3)

combined_cluster<-rbind(cluster1,cluster2,cluster3)

write.csv(combined_cluster, "combined_cluster2.csv",row.names=F,sep='\t')
```



## Differential expression of gene analysis:

### HYS vs iYS (p adjusted value=0.05)

```{r,fig.keep='all', warning=FALSE,message=FALSE}
pds<-dds

pds$condition<-relevel(pds$condition, ref="iYS")

pds<-DESeq(pds)

res = results(pds, contrast=c("condition","HYS","iYS"), alpha=0.05)

summary(res)

```


```{r,fig.keep='all', results='hide',warning=FALSE,message=FALSE}
sigs<-na.omit(res)

sigs.df<-as.data.frame(sigs)

columns(org.Hs.eg.db)

sigs.df$symbol<-mapIds(org.Hs.eg.db, keys=rownames(sigs.df), keytype = "ENSEMBL",column="SYMBOL")

sigs.df<-sigs.df%>%filter(!str_detect(symbol,'NA'))


```


Retrieve gene ontology terms associated with upregulated genes in HYS vs iYS samples. 

```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange >5,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```

Retrieve gene ontology terms associated with downregulated genes in HYS samples
```{r, results='hide',fig.keep='all', fig.width=10, fig.height=10, warning=FALSE,message=FALSE}
genes_to_test <- rownames(sigs[sigs$log2FoldChange < -2,])

GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")

as.data.frame(GO_results)


plot(barplot(GO_results, showCategory = 15))
```
